pkg/stanza/fileconsumer/file_threadpool.go

```diff
@@ -0,0 +1,265 @@
+		m.knownFiles[i].generation++
+	}
+	m.knownFilesLock.Unlock()
+	batchesProcessed := 0
+	// Get the list of paths on disk
+	matches, err := m.fileMatcher.MatchFiles()
+	if err != nil {
+		m.Errorf("error finding files: %s", err)
+	}
+	for len(matches) > m.maxBatchFiles {
+		m.consumeConcurrent(ctx, matches[:m.maxBatchFiles])
+
+		if m.maxBatches != 0 {
+			batchesProcessed++
+			if batchesProcessed >= m.maxBatches {
+				return
+			}
+		}
+
+		matches = matches[m.maxBatchFiles:]
+	}
+	m.consumeConcurrent(ctx, matches)
+
+	// Any new files that appear should be consumed entirely
+	m.readerFactory.fromBeginning = true
+	m.syncLastPollFilesConcurrent(ctx)
+}
+
+func (m *Manager) worker(ctx context.Context) {
+	defer m.workerWg.Done()
+	for {
+		select {
+		case wrapper, ok := <-m.readerChan:
+			if !ok {
+				return
+			}
+			r, fp := wrapper.reader, wrapper.trieKey
+			r.ReadToEnd(ctx)
+			if m.deleteAfterRead && r.eof {
+				r.Delete()
+			} else {
+				// Save off any files that were not fully read.
+				m.knownFilesLock.Lock()
+				m.knownFiles = append(m.knownFiles, r)
+				m.knownFilesLock.Unlock()
+			}
+			m.trieDelete(fp)
+		}
+	}
+}
+
+func (m *Manager) makeReaderConcurrent(filePath string) (*reader, *fingerprint.Fingerprint) {
+	fp, file := m.makeFingerprint(filePath)
+	if fp == nil {
+		return nil, nil
+	}
+
+	// check if the current file is already being consumed
+	if m.isCurrentlyConsuming(fp) {
+		if err := file.Close(); err != nil {
+			m.Errorf("problem closing file", "file", file.Name())
+		}
+		return nil, nil
+	}
+
+	// Exclude any empty fingerprints or duplicate fingerprints to avoid doubling up on copy-truncate files
+	if m.checkDuplicates(fp) {
+		if err := file.Close(); err != nil {
+			m.Errorf("problem closing file", "file", file.Name())
+		}
+		return nil, nil
+	}
+	m.currentFps = append(m.currentFps, fp)
+
+	reader, err := m.newReaderConcurrent(file, fp)
+	if err != nil {
+		m.Errorw("Failed to create reader", zap.Error(err))
+		return nil, nil
+	}
+	return reader, fp
+}
+
+func (m *Manager) consumeConcurrent(ctx context.Context, paths []string) {
+	m.Debug("Consuming files")
+	m.clearOldReadersConcurrent(ctx)
+	for _, path := range paths {
+		reader, fp := m.makeReaderConcurrent(path)
+		if reader != nil {
+			// add fingerprint to trie
+			m.triePut(fp)
+			m.readerChan <- readerEnvelope{reader: reader, trieKey: fp}
+		}
+	}
+	m.clearCurrentFingerprints()
+}
+
+func (m *Manager) isCurrentlyConsuming(fp *fingerprint.Fingerprint) bool {
+	m.trieLock.RLock()
+	defer m.trieLock.RUnlock()
+	return m.trie.HasKey(fp.FirstBytes)
+}
+
+func (m *Manager) triePut(fp *fingerprint.Fingerprint) {
+	m.trieLock.Lock()
+	defer m.trieLock.Unlock()
+	m.trie.Put(fp.FirstBytes)
+}
+
+func (m *Manager) trieDelete(fp *fingerprint.Fingerprint) {
+	m.trieLock.Lock()
+	defer m.trieLock.Unlock()
+	m.trie.Delete(fp.FirstBytes)
+}
+
+func (m *Manager) clearOldReadersConcurrent(ctx context.Context) {
+	m.knownFilesLock.Lock()
```

Comment:
@djaglowski I have made `clearOldReadersConcurrent()` a kind of blocking call based on the observations made while running benchmarks. 
We wait for all aggregated readers to finish reading before moving forward. 
 
Observations:

- If the number of files matched is significantly higher than `maxConcurrentFiles` then we are likely going to batch the files more than a couple of times.
- This causes `workers()` to overwhelm and they'll all be busy.
- If we try to send already finished readers to the `workers()` anyway, it will cause a bottleneck, and benchmarks show a significant delay in receiving logs.
- It's better to create separate goroutines to process finished readers and let the `workers()` loosen themselves up for the upcoming `poll()` cycle.
- In this way, we respect `max_concurrent_files`, `workers()` are happy, all's good. The benchmark performs as expected.